### DISCUSSION TITLE
Kekulization exception with SMARTS

### DISCUSSION THREAD

**Discussion by DavidACosgrove** [rdkit/rdkit#5514] [Q&A]
I don't know whether this is a bug or not, but it seems a bit odd to me.  It's in C++, for a change!  On the current master branch, built from source on macOS Monterey.

auto m1 = "c1ccccc1C-,=O"_smarts;

gives a molecule.

auto m1 = "c1ccccc1-C-,=O"_smarts;
throws a kekulization exception:

Can't kekulize mol.  Unkekulized atoms: 5

Is that expected?  I would have thought both should be ok.

**Comment by ptosco** [comment]
Hi Dave, this seems to work for me on current `master`:
```
from rdkit import Chem

Chem.MolFromSmarts("c1ccccc1C-,=O")
```
![image](https://user-images.githubusercontent.com/5244385/186361989-fdc0fde8-7ad8-4c50-ba7d-23b8725be252.png)

```
Chem.MolFromSmarts("c1ccccc1-C-,=O")
```
![image](https://user-images.githubusercontent.com/5244385/186362081-6bb3775c-4317-44f6-8545-f193cf3df814.png)

**Comment by DavidACosgrove** [comment]
Thanks Paulo.  Further investigation shows that it's in prepareMolForDrawing:
```
  std::string nameBase("testGithub5511_");
  auto s1 = "c1ccccc1C-,=O"_smarts;
  MolDraw2DUtils::prepareMolForDrawing(*s1);
  auto s2 = "c1ccccc1-C-,=O"_smarts;
  MolDraw2DUtils::prepareMolForDrawing(*s2);
```
throws the exception on the second call to prepareMolForDrawing.  But that is also a bit odd since I assume your pictures went down a similar route.

**Reply by ptosco** [reply]
@DavidACosgrove I do get the exception if I call `prepareMolForDrawing`:
```python
rdMolDraw2D.PrepareMolForDrawing(Chem.MolFromSmarts("c1ccccc1-C-,=O"))
[11:59:56] Can't kekulize mol.  Unkekulized atoms: 5
---------------------------------------------------------------------------
KekulizeException                         Traceback (most recent call last)
/tmp/ipykernel_232767/4056974730.py in <module>
----> 1 rdMolDraw2D.PrepareMolForDrawing(Chem.MolFromSmarts("c1ccccc1-C-,=O"))

KekulizeException: Can't kekulize mol.  Unkekulized atoms: 5
```
The reason why I don't get it in Jupyter is that the code in `rdkit/Chem/Draw/__init__.py` checks if the molecule can be kekulized and if not it calls `rdMolDraw2D.PrepareMolForDrawing` with `kekulize=False`.
You can indeed see in the picture above that the ring is not kekulized. In general query molecules are difficult beasts to kekulize, so I am not too surprised it fails, even though you in principle it should not.

**Reply by DavidACosgrove** [reply]
I can understand that queries can be difficult to kekulize, but I would have thought the second query, where the cC bond is explicitly single, would be easier to deal with than the first case where it's single or aromatic.

**Reply by ptosco** [reply]
I think I know what happens. In this molecule
```python
m = Chem.MolFromSmarts("c1ccccc1-C-,=O")
m
```
![image](https://user-images.githubusercontent.com/5244385/186395519-eaeaf7bf-e78b-424d-bcd5-36e4fafa9d1a.png)
all aromatic ring bonds are labelled as query bonds
```python
for b in m.GetBonds():
    print(b.GetIdx(), b.GetIsAromatic(), b.HasQuery())
0 True True
1 True True
2 True True
3 True True
4 True True
5 False True
6 False True
7 True True
```
as they are not explicitly marked as aromatic, hence kekulization may fail, which is why the code in `rdkit/Chem/Draw/__init__.py` does exactly that check:
```python
def _okToKekulizeMol(mol, kekulize):
  if kekulize:
    for bond in mol.GetBonds():
      if bond.GetIsAromatic() and bond.HasQuery():
        return False
    return True
  return kekulize
```
If you explicitly mark them as aromatic, then `PrepareMolDrawing` won't fail, because they are not query bonds anymore:
```python
m = Chem.MolFromSmarts("c1:c:c:c:c:c:1-C-,=O")
m
```
![image](https://user-images.githubusercontent.com/5244385/186396325-bfe3f94e-1401-4d8d-8130-b35fabb5674b.png)
```python
for b in m.GetBonds():
    print(b.GetIdx(), b.GetIsAromatic(), b.HasQuery())
0 False True
1 False True
2 False True
3 False True
4 False True
5 False True
6 False True
7 True True
```

**Reply by ricrogz** [reply]
I think this is the same issue I reported in https://github.com/rdkit/rdkit/issues/5505 last week. The issue, I think, is that Query bonds (as the "single-or-double" one in your example) cause the end atoms to report valence 0, which I think is the cause the Kekulization doesn't work as expected.
